Dm 3111

bradjohnson92008 opened the pull request:
### JIRA issue link
https://agile6.atlassian.net/browse/DM-3111

## Description - what does this code do?
This code adds Clinincal Resource Hubs (CRHs) to the facility Index page "Faciliities" Combo Box and adds them to the list of facilities.  CRHs are displayed last in both cases.
CRH show page - see figma for design (https://www.figma.com/file/ccwLbt94U9QGfnuFCwOXzY/Final-designs---VA-Diffusion-Marketplace?node-id=103%3A69).  Note** the summary text at the top i.e., number of innovations created and number of adoptions are NOT hyperlinks.  Displays states that the CRH covers as well as  the number of created innovations by the CRH as well as the number of adoptions implemented by the CRH.  



## Testing done - how did you test it/steps on how can another person can test it 
1.  Log in to dm and navigate to /facilities
2. Verify all CRHs are displayed in the "Facilities Combox list filter" at the bottom of the list and in visn # order.
![image](https://user-images.githubusercontent.com/60527788/158700176-fd5078f7-4558-4d85-a265-69c093c3fa88.png)
3.  Scroll to the bottom of the facility list and verify the CRHs are displayed at the bottom in Visn # order.
![image](https://user-images.githubusercontent.com/60527788/158700369-c992c2b9-1c56-4a6b-a630-bdcdd58a9932.png)
4.  Select any CRH facility name link and verify the url is /crh/{VISN #} and links to the CRH show page for that CRH.
![image](https://user-images.githubusercontent.com/60527788/158700627-300b0b80-340a-43f5-9735-da63b054411b.png)
5.  Select any CRH from the "Facilities Combox List filter" and verify the CRH is filtered and displays.
![image](https://user-images.githubusercontent.com/60527788/158700891-ce9ec0e3-338b-4924-a1ad-c04c1cf5cab5.png)
6.  Verify the filtered CRH Name link takes you to the crh show page for that CRH.
7. CRH show page - test that the number of innovations created and adopted are correct.
8. Verify the Categories associated with the created innovations and adopted innovations are correct and that the Categories filter works in both scenarios.
9. Verify the "Sort by" filter works in both scenarios as well.
10. Verify that the load more functionality works as intended.  (loading 6 cards at a time..)

## Screenshots, Gifs, Videos from application (if applicable)
crh-show page search section
![image](https://user-images.githubusercontent.com/60527788/158699441-d3f57f6d-feea-43b8-b7a7-6c6ec77e5170.png)


## Link to mock-ups/mock ups (image file if you have it) (if applicable)


## Acceptance criteria

- Add CRH to Facilities List on index page
- Create CRH Show Page 
- [https://www.figma.com/file/ccwLbt94U9QGfnuFCwOXzY/Final-designs---VA-Diffusion-Marketplace?node-id=103%3A69 - Connect to preview](https://www.figma.com/file/ccwLbt94U9QGfnuFCwOXzY/Final-designs---VA-Diffusion-Marketplace?node-id=103%3A69) 
- List them in facilities list at the end
- List CRHs in alpha/numerical order
- Different route to CRH show page (/crh/#)
- display 6 practice cards at a time on show page
- CHR’s should not display at all if the “complexity” filter is being used.
- If VISN filter is selected (and no complexity filter selected) only the CHR for the selected VISN should display (at the bottom).

## Definition of done
- [ ] Unit tests written (if applicable)
- [ ] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating JIRA issue
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

emyl3 commented on app/views/va_facilities/_index_table_row.html.erb:
```diff
@@ -1,10 +1,24 @@
+    <th scope="row" role="rowheader" class="grid-col-4"><a class="usa-link" href='<%= path %>'><%= facility["official_station_name"] %>
+      <% if facility.class.name != "ClinicalResourceHub"  %> (<%= facility["common_name"] %>)<% end %></a>
+    </th>
+    <td class="grid-col-1"><%= facility.class.name === "ClinicalResourceHub" ? facility.visn.state : facility.street_address_state %></td>
```
I thought we weren't assigning one state to a CRH

bradjohnson92008 replied:
Yeah... sorry I was using visn.state field for this... but since they are associated with multiple states... just displaying '-' for CRH state column... also displaying "-" for complexity field... was displaying "N/A"..  lol

emyl3 commented on app/views/va_facilities/_index_table_row.html.erb:
```diff
@@ -1,10 +1,24 @@
+    <th scope="row" role="rowheader" class="grid-col-4"><a class="usa-link" href='<%= path %>'><%= facility["official_station_name"] %>
+      <% if facility.class.name != "ClinicalResourceHub"  %> (<%= facility["common_name"] %>)<% end %></a>
+    </th>
+    <td class="grid-col-1"><%= facility.class.name === "ClinicalResourceHub" ? facility.visn.state : facility.street_address_state %></td>
```
I thought we weren't assigning one state to a CRH

bradjohnson92008 replied:
Yeah... sorry I was using visn.state field for this... but since they are associated with multiple states... just displaying '-' for CRH state column... also displaying "-" for complexity field... was displaying "N/A"..  lol

emyl3 commented on spec/features/crh_spec.rb:
```diff
@@ -14,6 +14,12 @@
         expect(page).to_not have_content('We\'re sorry, we can\'t find the page you\'re looking for. It might have been removed, changed names, or is otherwise unavailable.')
       end
     end
+    describe 'crh show page works' do
```
we are adding lots of functionality in this PR can we add tests to cover these scenarios...
some suggestions on tests to add:

- in the `va_facility_spec.rb` adding tests to ensure the CRHs are being displayed correctly

- in this `crh_spec.rb` adding tests to ensure the content displays correctly, tooltip works (needs to be added), search and filtering sorting works as intended

- in the `breadcrumbs_spec.rb` ensuring the breadcrumbs appear correctly and has the appropriate link for this new page

bradjohnson92008 replied:
Added tests for these cases..

emyl3 commented on spec/features/crh_spec.rb:
```diff
@@ -14,6 +14,12 @@
         expect(page).to_not have_content('We\'re sorry, we can\'t find the page you\'re looking for. It might have been removed, changed names, or is otherwise unavailable.')
       end
     end
+    describe 'crh show page works' do
```
we are adding lots of functionality in this PR can we add tests to cover these scenarios...
some suggestions on tests to add:

- in the `va_facility_spec.rb` adding tests to ensure the CRHs are being displayed correctly

- in this `crh_spec.rb` adding tests to ensure the content displays correctly, tooltip works (needs to be added), search and filtering sorting works as intended

- in the `breadcrumbs_spec.rb` ensuring the breadcrumbs appear correctly and has the appropriate link for this new page

bradjohnson92008 replied:
dup

emyl3 commented on app/controllers/va_facilities_controller.rb:
```diff
@@ -1,16 +1,35 @@
   include PracticeUtils
   before_action :set_va_facility, only: [:show, :created_practices, :update_practices_adopted_at_facility]
-
+  helper_method :crh_created_practices_count, :crh_adopted_practices_count
```
can we make a `va_facilities_helper.rb` file and move this and its associated methods out of the controller

bradjohnson92008 replied:
Yep - moved methods to va_facilities_helper.rb file.

emyl3 commented on app/views/clinical_resource_hubs/show.html.erb:
```diff
@@ -1,8 +1,34 @@
+  <%= javascript_include_tag 'clinical_resource_hubs/crh_show', 'data-turbolinks-track': 'reload' %>
+<% end %>
+
+<div id="crh-show" class="grid-col-12" data-reload="true">
```
the style on this page is slightly off from the Figma design

Figma
<img width="940" alt="Screen Shot 2022-03-21 at 12 08 01 PM" src="https://user-images.githubusercontent.com/20211771/159333647-0fc03a38-63f2-44b1-ae0d-677cb929609d.png">

This branch
<img width="1916" alt="Screen Shot 2022-03-21 at 12 50 07 PM" src="https://user-images.githubusercontent.com/20211771/159333797-32a181b7-a472-48be-b0d6-d74eb35cb9b2.png">
   
Things I am noticing:

- no definition underline for "Clinical Resource Hub"
- not enough padding for breadcrumbs
- different font for text below heading
-

bradjohnson92008 replied:
Thx for your help on this one Elisa.

emyl3 commented on app/views/clinical_resource_hubs/show.html.erb:
```diff
@@ -1,8 +1,34 @@
+  <%= javascript_include_tag 'clinical_resource_hubs/crh_show', 'data-turbolinks-track': 'reload' %>
+<% end %>
+
+<div id="crh-show" class="grid-col-12" data-reload="true">
```
the style on this page is slightly off from the Figma design

Figma
<img width="940" alt="Screen Shot 2022-03-21 at 12 08 01 PM" src="https://user-images.githubusercontent.com/20211771/159333647-0fc03a38-63f2-44b1-ae0d-677cb929609d.png">

This branch
<img width="1916" alt="Screen Shot 2022-03-21 at 12 50 07 PM" src="https://user-images.githubusercontent.com/20211771/159333797-32a181b7-a472-48be-b0d6-d74eb35cb9b2.png">
   
Things I am noticing:

- no definition underline for "Clinical Resource Hub"
- not enough padding for breadcrumbs
- different font for text below heading
- extra space in between categories and select
- the results count needs to be a little lower

bradjohnson92008 replied:
dup

emyl3 reviewed (changes requested):
Left an initial review. Good work so far!

JoshingYou1 commented on app/controllers/clinical_resource_hubs_controller.rb:
```diff
@@ -1,9 +1,60 @@
   def show
+    visn_id = @visn.id
+    visn_number = @visn.number
+    @primary_visn_liaison = VisnLiaison.find_by(visn: @visn, primary: true)
```
Do we need this? I don't think we're using it anywhere.

bradjohnson92008 replied:
Good catch... ur right we don't need those lines.

JoshingYou1 commented on app/controllers/va_facilities_controller.rb:
```diff
@@ -1,16 +1,35 @@
+
   def load_facilities_index_rows
-    if params[:facility].present?
+    if params[:facility].present? && params[:facility].to_i > 0
```
Do we need this second check? Is there an instance where the facility param is `0`?

bradjohnson92008 replied:
Nope we don't need it... good catch.

JoshingYou1 commented on app/controllers/va_facilities_controller.rb:
```diff
@@ -1,16 +1,35 @@
       @facilities = [VaFacility.cached_va_facilities.order_by_station_name.includes([:visn]).find(params[:facility])]
+
+    elsif params[:crh].present? && params[:crh].to_i > 0
+      @facilities = [ClinicalResourceHub.find_by_id(params[:crh].to_i)]
```
Same question as above

bradjohnson92008 replied:
Correct - we don't need that additional check.

JoshingYou1 commented on app/models/practice.rb:
```diff
@@ -375,6 +379,14 @@ def self.search_practices(search_term = nil, sort = 'a_to_z', categories = nil,
+  #   debugger
+  #   practices = search_practices(search_term, sort, categories, is_user_guest)
+  #   practices.select { |pr| pr.practice_origin_facilities.pluck(:clinical_resource_hub_id).include?(crh_id) }
+  # end
```
We can remove this if we're not using it.

bradjohnson92008 replied:
Removed.

JoshingYou1 commented on app/views/clinical_resource_hubs/show.html.erb:
```diff
@@ -1,8 +1,34 @@
+        <%= @crh.official_station_name %>
+      </h1>
+      <div class="grid-row grid-gap-3 margin-bottom-10">
+        <div class="grid-col-12 line-height-26"> This Clinical Resource Hub (CRH) serves veterans in <%= @states_str %>. It has created <%= @practices_created_by_crh.count %> innovations and has adopted <%= @practices_adopted_by_crh.count %> innovations.
```
We aren't taking into account a single origin facility/adoption here. If the CRH has just one adoption, it reads "adopted 1 innovations."

bradjohnson92008 replied:
Got it - good catch.

JoshingYou1 reviewed (changes requested):
Nice work on this Brad. I added some comments. I noticed some of the styling does not reflect the Figma mockup, so I'll add them here:

1. No modal for the CRH title
_**DM-3111**_
![](https://user-images.githubusercontent.com/34111449/159342594-fc314c36-932c-4908-9b4f-ac2452d24cdd.png)
_**Figma Mockup**_
![](https://user-images.githubusercontent.com/34111449/159342737-6af7da98-8d9f-4aa4-a806-bd3beea24bb3.png)

2. No links for the adoption/origin counts below title:
_**DM-3111**_
![](https://user-images.githubusercontent.com/34111449/159342886-dc6880d7-f98c-47ac-acbc-2a2c917c25da.png)
_**Figma Mockup**_
![](https://user-images.githubusercontent.com/34111449/159342942-6758c087-b382-4ad6-87e4-43d70453065a.png)

3. Styling issue on category combobox when active
![](https://user-images.githubusercontent.com/34111449/159343036-ee6138be-3b0e-4350-8d02-1119eb0ac7ce.png)

4. Search input is too small
_**DM-3111**_
![](https://user-images.githubusercontent.com/34111449/159343241-6466234f-7316-4568-946d-826f2a7443a3.png)
_**Figma Mockup**_
![](https://user-images.githubusercontent.com/34111449/159343347-0df96a70-c40e-4d97-ae9e-80c32e343e3e.png)

bradjohnson92008 commented on app/controllers/va_facilities_controller.rb:
```diff
@@ -1,16 +1,35 @@
   include PracticeUtils
   before_action :set_va_facility, only: [:show, :created_practices, :update_practices_adopted_at_facility]
-
+  helper_method :crh_created_practices_count, :crh_adopted_practices_count
```
dup

bradjohnson92008 commented:
Josh - I added the modal for CRH info.  Summary text at top of page for innovations adopted and created are NOT links.  Fixed the styling for the Categories Combo box and increased the height for the search input in crh.scss to match.

bradjohnson92008 commented on app/helpers/practices_helper.rb:
```diff
@@ -18,6 +18,57 @@ def options_for_states
     state_hash_str
   end
 
+  def states_lookup
```
Yep - refactored this and removed redundant code.  Good Catch

emyl3 commented on app/models/clinical_resource_hub.rb:
```diff
@@ -39,4 +39,27 @@ def self.cached_clinical_resource_hubs
+    DiffusionHistory.where(clinical_resource_hub_id: id).count
+  end
+
+  def get_crh_adopted_practices( crh_id, options = { :is_user_guest => true })
```
let's avoid using the rockets `=>` when we can instead we can do `is_user_guest: true` same comment for other parts we are using `=>`

bradjohnson92008 replied:
Got it.

emyl3 commented on app/models/clinical_resource_hub.rb:
```diff
@@ -39,4 +39,27 @@ def self.cached_clinical_resource_hubs
     end
   end
+
+  def practices_created_by_crh_count
```
are we using this?

bradjohnson92008 replied:
nope... i removed it.

emyl3 commented on app/models/clinical_resource_hub.rb:
```diff
@@ -39,4 +39,27 @@ def self.cached_clinical_resource_hubs
+    PracticeOriginFacility.where(clinical_resource_hub_id: id).count
+  end
+
+  def practices_adopted_by_crh_count
```
are we using this?

bradjohnson92008 replied:
Nope.. :)  removed.

emyl3 commented on app/models/practice.rb:
```diff
@@ -194,8 +194,12 @@ def highlight_attachment_s3_presigned_url(style = nil)
   scope :sort_by_retired, -> { order("retired asc") }
   scope :get_by_adopted_facility, -> (facility_id) { left_outer_joins(:diffusion_histories).where(diffusion_histories: {va_facility_id: facility_id}).uniq }
   scope :get_by_adopted_facility_and_crh, -> (facility_id, crh_id) { left_outer_joins(:diffusion_histories).where(diffusion_histories: {va_facility_id: facility_id}).or(left_outer_joins(:diffusion_histories).where(diffusion_histories: {clinical_resource_hub_id: crh_id})).uniq }
+  scope :get_by_adopted_crh, -> (crh_id) { (left_outer_joins(:diffusion_histories).where(diffusion_histories: {clinical_resource_hub_id: crh_id})).uniq }
```
I don't think we need tho wrap `left_outer_joins(:diffusion_histories).where(diffusion_histories: {clinical_resource_hub_id: crh_id})` with parentheses

bradjohnson92008 replied:
I think your right.. I removed the parentheses...

emyl3 commented on app/views/clinical_resource_hubs/_crh_categories_combo_box.html.erb:
```diff
@@ -0,0 +1,12 @@
+<div class="usa-combo-box grid-col-4 practice-<%= practice_type %>-categories-combo-box" data-placeholder="All categories">
+  <label for="crh-search-<%= practice_type %>-categories" class="line-height-26">Categories</label>
+  <select id="crh-search-<%= practice_type %>-categories" class="usa-select" name="crh-search-<%= practice_type %>-categories">
+    <% if practice_categories.count > 0 %>
```
can we wrap the if statement around the entire div and add `practice_categories.present? && count check`

bradjohnson92008 replied:
done.

emyl3 commented on spec/features/crh_spec.rb:
```diff
@@ -14,6 +33,34 @@
+      end
+      it 'should display CRH info modal and close' do
+        visit '/crh/1'
+        # find('#crh_info_modal').click
```
we can remove this comment

bradjohnson92008 replied:
Done.

emyl3 reviewed (changes requested):
Left some additional comments, Brad!

bradjohnson92008 commented on app/helpers/practices_helper.rb:
```diff
@@ -18,6 +18,57 @@ def options_for_states
     state_hash_str
   end
 
+  def states_lookup
```
This was a duplicate - addressed in last round of fixes.

bradjohnson92008 commented on app/helpers/practices_helper.rb:
```diff
@@ -18,6 +18,20 @@ def options_for_states
     state_hash_str
   end
 
+  def states_lookup
```
Correct - I removed it.

bradjohnson92008 commented on app/controllers/clinical_resource_hubs_controller.rb:
```diff
@@ -1,12 +1,57 @@
+    @visn_va_facilities = VaFacility.get_by_visn(@visn).get_relevant_attributes
+    @visn_crh = ClinicalResourceHub.cached_clinical_resource_hubs.find_by(visn: @visn)
 
+    @practices_created_by_crh = helpers.is_user_a_guest? ? @crh.get_crh_created_practices(@crh.id, :is_user_guest => true) :
```
yep - much cleaner

bradjohnson92008 commented on app/controllers/clinical_resource_hubs_controller.rb:
```diff
@@ -1,12 +1,57 @@
+
+    @practices_created_json = practices_json(@practices_created_by_crh) unless @practices_created_by_crh == nil
+
+    @practices_adopted_by_crh = helpers.is_user_a_guest? ? @crh.get_crh_adopted_practices(@crh.id,  :is_user_guest => true) :
```
Got it.

bradjohnson92008 commented on app/controllers/clinical_resource_hubs_controller.rb:
```diff
@@ -1,12 +1,57 @@
+    sort_option = params[:sort_option] || 'a_to_z'
+    search_term = params[:search_term] ? params[:search_term].downcase : nil
+    categories = params[:categories] || nil
+    created_practices = helpers.is_user_a_guest? ? @crh.get_crh_created_innovations(@crh.id, search_term, sort_option, categories, true) : @crh.get_crh_created_innovations(@crh.id, search_term, sort_option, categories, false)
```
Got it.

bradjohnson92008 commented on app/assets/javascripts/clinical_resource_hubs/crh_show.es6:
```diff
@@ -0,0 +1,30 @@
+            $(table).removeClass("display-none");
+            $(loadingSpinner).addClass("display-none");
+            $(crhShow).data('reload', false);
+            //alert(data);
```
got it.

bradjohnson92008 commented on app/controllers/clinical_resource_hubs_controller.rb:
```diff
@@ -1,12 +1,57 @@
   private
   def set_crh
     @visn = params[:id].present? ? Visn.find_by!(number: params[:id]) : Visn.find_by!(number: params[:number])
+    @visn_obj = Visn.find_by!(number: params[:number])
```
Yes... appears to be working fine with the original @visn prop... I forget why I thought i had to do that.... lol  anyway I tested the page pretty thoroughly.. signed in - not signed in ... etc.

JoshingYou1 reviewed (changes requested):
Nice work Brad! I just noticed a couple of things:

1. If a CRH has 0 created/adopted facilities, the search bar styling is a little wonky.
![](https://user-images.githubusercontent.com/34111449/160447378-1a751fa3-887d-4fa2-b6fe-e9fc4e1f15df.png)

2. Should the VISN select only bring back the CRH that belongs to that VISN? Currently, it brings back the whole list.
![](https://media.giphy.com/media/BY3vDFyAYvZLYjaezZ/giphy.gif)

bradjohnson92008 commented on app/controllers/clinical_resource_hubs_controller.rb:
```diff
@@ -1,7 +1,49 @@
-  before_action :set_crh, only: [:show]
+  include PracticesHelper
+  include PracticeUtils
+  include StatesHelper
```
Got it.

bradjohnson92008 commented:
I updated the AC for not displaying CRH's when the 'complexity' filter is selected.. and for only showing the CHR for the selected VISN if the VISN filter is selected...  (given that the complexity filter is not selected).  Also - best solution i could find for the Created or Adopted radio buttons having 0 matches is to just disable the radio so it can't be selected...  if both are 0 then the search section does not display at all.  Let me know if that works for you.  thx

JoshingYou1 reviewed (commented):
Thank you for fixing up those two things Brad! I just left a few nitpick comments. Awesome job! 😄

bradjohnson92008 commented on app/controllers/va_facilities_controller.rb:
```diff
@@ -1,30 +1,36 @@
       end
-
       if params[:complexity].present?
+        @clinical_resource_hubs = []
```
This is to clear all the clinical_resource_hubs... since they are not applicable when filtering by "complexity".  I am adding the the CRH array to the @facilities list on line 31..  so we only have to do this once for all the search filter scenarios.  fyi - I also added the cached_clinical_resource_hubs scope on line 20.

bradjohnson92008 commented on app/controllers/va_facilities_controller.rb:
```diff
@@ -1,30 +1,36 @@
       if params[:visn].present?
         @facilities = @facilities.where(visns: { number: params[:visn] })
+        visn_id = Visn.where(number: params["visn"].to_i).first().id
+        @clinical_resource_hubs = @clinical_resource_hubs.where(visn_id: visn_id).sort_by(&:id)
```
This is the clinical_resource_hubs array that gets populated here (correct only 1 when there is a VISN filter) and then added to the faciliaites list array on line 31.

bradjohnson92008 replied:
true Elisa... I removed the sort_by when filtering by visn for crh's

emyl3 reviewed (commented):
Just the couple of minor things Josh pointed out and we should be good! Thanks, Brad!

bradjohnson92008 commented on app/controllers/va_facilities_controller.rb:
```diff
@@ -1,30 +1,36 @@
   def index
-    @facilities = VaFacility.cached_va_facilities.select(:common_name, :id, :official_station_name).order_by_station_name
+    @facilities = VaFacility.cached_va_facilities.select(:common_name, :id, :visn_id, :official_station_name).order_by_station_name.includes([:visn])
+    @clinical_resource_hubs = ClinicalResourceHub.all
```
Got it - I added it on line 20 as well - in the load_facility_index_rows method.  Good catch.

JoshingYou1 reviewed (approved):
LGTM! Nice work Brad!

emyl3 reviewed (approved):
Great work on this! 😸
